sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala

```diff
@@ -97,7 +99,42 @@ private[hive] trait SaveAsHiveFile extends DataWritingCommand {
       options = Map.empty)
   }
 
-  protected def getExternalTmpPath(
+  // Mostly copied from Context.java#getMRTmpPath of Hive 2.3.
+  // Visible for testing.
+  private[execution] def getNonBlobTmpPath(
+      hadoopConf: Configuration,
+      sessionScratchDir: String,
+      scratchDir: String): Path = {
+
+    // Hive's getMRTmpPath uses nonLocalScratchPath + '-mr-10000',
+    // which is ruled by 'hive.exec.scratchdir' including file system.
+    // This is the same as Spark's #oldVersionExternalTempPath.
+    // Only difference between #oldVersionExternalTempPath and Hive 2.3.0's is HIVE-7090.
+    // HIVE-7090 added user_name/session_id on top of 'hive.exec.scratchdir'
+    // Here it uses session_path unless it's emtpy, otherwise uses scratchDir.
+    val sessionPath = if (!sessionScratchDir.isEmpty) sessionScratchDir else scratchDir
+    val mrScratchDir = oldVersionExternalTempPath(new Path(sessionPath), hadoopConf, sessionPath)
+    logDebug(s"MR scratch dir '$mrScratchDir/-mr-10000' is used")
+    val path = new Path(mrScratchDir, "-mr-10000")
+    val scheme = Option(path.toUri.getScheme).getOrElse("")
+    if (scheme.equals("file")) {
+      logWarning("Temporary data will be written into a local file system " +
+        "(scheme: '$scheme', path: '$mrScratchDir'). If your Spark is not in local mode, " +
```

Comment:
s"" for string interpolation.

Reply:
Oh it was my bad. Reverted only the necessary part.